### PR TITLE
Add tutorial for DPVS config and update README.md

### PR DESCRIPTION
**Pull request by beacer**
1. add tutorial for config: `FNAT`/`SNAT`/`DR`, `keepalived`/`ospfd`, virtual devices.
2. update `README.md`

### CODE REVIEW

File: doc/tutorial.md [hunk 1]
```diff
@@ -0,0 +1,594 @@
...
+
+<a id='term'/>
+
+# Terminology
+
+For the meanings of *Full-NAT* (`FNAT`), `DR`, `toa`, `OSPF`/`ECMP` and `keepalived`, pls refer [LVS](www.linuxvirtualserver.org) and [Alibaba/LVS](https://github.com/alibaba/LVS/tree/master/docs).
+
+Note `DPVS` support `FNAT`, `DR`, `SNAT` forwarding modes, and each mode can be configured as `one-arm` or `two-arm` topology, with or without `ospfd`/`keepalived`. There're too many combinations, I cannot list all the examples here. Let's just give some popular working models used in our daily work.
+
+> `Tunneling` and `SNAT` is not supported because we didn't see strong demand yet, at least from *iQiYi*. May add them in the future. :)
```
Comment (ywc689) [inline doc/tutorial.md]: `NAT` not support,  not `SNAT`

File: doc/tutorial.md [hunk 2]
```diff
@@ -0,0 +1,594 @@
+DPVS Tutorial
+=============
+
+* [Terminology](#term)
+  - [*One-arm* and *two-arm*](#one-two-arm)
```
Comment (ywc689) [inline doc/tutorial.md]: *two-arms*?

Comment (beacer) [reply doc/tutorial.md]: two-arm, no `s`. refer http://networkgeekstuff.com/networking/basic-load-balancer-scenarios-explained/.

File: doc/tutorial.md (hunk 1 above)
Comment (beacer) [reply doc/tutorial.md]: Ok

File: doc/tutorial.md [hunk 3]
```diff
@@ -0,0 +1,594 @@
...
+
+For *one-arm*, only one DPDK intreface needed, and you can refer `conf/dpvs.conf.single-nic.sample`.
+
+<a id='kni'/>
+
+## KNI Device
+
+Like `LVS`, `DPVS` can be deployed as different sort of *Cluster* models for High-Available (HA) purpose. Both *OSPF/ECMP* and *Master/Backup* models are supported. *OSPF/ECMP* model need package `quagga` and its `zebra` and `ospfd` programs. And *master/back* model need `Keepalived`.
+
+Consider `DPDK` application manages the networking interface completely (except the extra control NIC if exist), Linux Kernel and programs run on Kernel TCP/IP stack cannot receive packets from `DPDK` interface directly. To make Linux programs like `sshd`, `zebra/ospfd` and `keepalived` work, DPDK `kni` device is used. Then the Linux programs can working on `kni` device with Linux TCP/IP stack. Actually, `DPVS` pass the packets it's not interested in to `kni` device, like `OSPF/VRRP/ssh` packets. So that the programs "listening" on Linux stack are able to handle them.
```
Comment (ywc689) [inline doc/tutorial.md]: Considering ...

Comment (ywc689) [reply doc/tutorial.md]: `DPVS` pass the packets to `kni` device  which it's not interested in, like `OSPF/VRRP/ssh` packets. So the programs ...

File: doc/tutorial.md [hunk 4]
```diff
@@ -0,0 +1,594 @@
...
+
+<a id='fnat'/>
+
+# Full-NAT Mode
+
+<a id='simple-fnat'/>
+
+## Simple Full-NAT (two-arm)
+
+This is a simple example for FullNAT (`FNAT`), forwarding between two interfaces. Let's say one is WAN interface (`dpdk1`) and another is LAN interface (`dpdk0`).
```
Comment (ywc689) [inline doc/tutorial.md]: Assume one is WAN ...

File: doc/tutorial.md [hunk 5]
```diff
@@ -0,0 +1,594 @@
...
+$ ./ipvsadm  -G
+VIP:VPORT            TOTAL    SNAT_IP              CONFLICTS  CONNS
+10.0.0.100:80        1
+                              192.168.100.200      0          0
+```
+
+And now have a try if FNAT (two-arm) works. I've set the Nginx server on RS (with TOA module) to response the real Client IP/Port directly as plain text (not html).
+
+```bash
+client$ curl 10.0.0.0.100
```
Comment (ywc689) [inline doc/tutorial.md]: client$ curl 10.0.0.100:80

File: doc/tutorial.md [hunk 6]
```diff
@@ -0,0 +1,594 @@
...
+```
+
+```bash
+$ ./ipvsadm  -G
+VIP:VPORT            TOTAL    SNAT_IP              CONFLICTS  CONNS
+10.0.0.100:80        1
+                              192.168.100.200      0          0
+```
+
+And now have a try if FNAT (two-arm) works. I've set the Nginx server on RS (with TOA module) to response the real Client IP/Port directly as plain text (not html).
```
Comment (ywc689) [inline doc/tutorial.md]: And now verify if FNAT (two-arms) works. I've setup the Nginx server on RS (with TOA module) to response HTTP request of 10.0.0.100:80 as plain text (not html).

File: doc/tutorial.md [hunk 7]
```diff
@@ -0,0 +1,594 @@
...
+<a id='lip-routes-toa'/>
+
+##### Something about *LIP*, *routes* and *TOA*
+
+`LIP` or *Local-IP* is needed for FNAT translation, clients' *CIP:cport* will be replaced with *LIP:lport*, while *VIP:vport* will be translated to RS's *RIP:rport*. That's why the mode called "Full-NAT" I think.
+
+Pls use `ipvsadm --add-laddr` to set `LIP` instead of `dpip addr add ...`. Because the both *ipvs* and *inet* module need `LIP` address, and *sapool* option will be set automatically by `ipvsadm --add-laddr`. Or you have to use two commands,
+
+```bash
+./ipvsadm --add-laddr -z <LIP> ...
```
Comment (ywc689) [inline doc/tutorial.md]: feel confusing with above-mentioned

File: doc/tutorial.md [hunk 8]
```diff
@@ -0,0 +1,594 @@
...
+
+## Full-NAT with OSPF/ECMP (two-arm)
+
+To work with *OSPF*, the patch in `patch/dpdk-16.07/` must be applied to *DPDK-16.07* and the correct `rte_kni.ko` should be installed.
+
+`DPVS` OSPF-cluster model looks like this, it leverage `OSPF/ECMP` for HA and high-scalability. This model is widely used in practice.
+
+![fnat-ospf-two-arm](pics/fnat-ospf-two-arm.png)
+
+For `DPVS`, things become more complicated. As mentioned above, `DPDK` program (here is `dpvs`) have fully control of DPDK NICs, so Linux program (`ospfd`) need receive/send packets through `kni` device (`dpdk1.kni`) related to DPDK device (`dpdk1`).
```
Comment (ywc689) [inline doc/tutorial.md]: fully control ->full control
need -> needs

File: doc/tutorial.md [hunk 9]
```diff
@@ -0,0 +1,594 @@
...
+
+> DPDK apps based on whole TCP/IP stack like user-space Linux/BSD has no this kind of configuration complexity, but more developing efforts are needed to porting `ospfd` and `keepalived` to the TCP/IP stack used by DPDK. Anyway, that's another solution.
```
Comment (ywc689) [inline doc/tutorial.md]: has no this kind of -> do not have this kind of

File: doc/tutorial.md [hunk 10]
```diff
@@ -0,0 +1,594 @@
...
+Thus, the internal relationship among interfaces and programs looks like below,
+
+![fnat-ospfd-kni.png](pics/fnat-ospfd-kni.png)
+
+Now the configuration has two parts, one is for `dpvs` and another is for `zebra/ospfd`.
+
+`dpvs` part is almost same with the example in [simple fnat](#simple-fnat), except
+
+* one more route to **kni-host** is needed to pass the packets received from `dpvs` device to Linux `kni` device.
+* VIP should not set to `dpvs` by `dpdp addr`, and need be set to `kni` instead, so that `ospfd` can aware and publish it.
```
Comment (ywc689) [inline doc/tutorial.md]: `dpdp addr` -> `dpip addr`
can aware -> can be aware of

File: doc/tutorial.md [hunk 11]
```diff
@@ -0,0 +1,594 @@
...
+#!/bin/sh -
+
+# routes for LAN access
+./dpip route add 192.168.100.0/24 dev dpdk0
+
+# add service <VIP:vport> to forwarding, scheduling mode is RR.
+# use ipvsadm --help for more info.
+./ipvsadm -A -t 123.1.2.3:80 -s rr
+
+# add two RS for service, forwarding mode is FNAT (-b)
```
Comment (ywc689) [inline doc/tutorial.md]: two RS -> two RSs

File: doc/tutorial.md [hunk 12]
```diff
@@ -0,0 +1,594 @@
...
+./ipvsadm -a -t 123.1.2.3:80 -r 192.168.100.2 -b
+./ipvsadm -a -t 123.1.2.3:80 -r 192.168.100.3 -b
+
+# add at Local-IP (LIP) for FNAT on LAN interface
```
Comment (ywc689) [inline doc/tutorial.md]: add at Local-IP (LIP) -> add Local-IPs (LIPs)

File: doc/tutorial.md [hunk 13]
```diff
@@ -0,0 +1,594 @@
...
+./ipvsadm --add-laddr -z 192.168.100.200 -t 123.1.2.3:80 -F dpdk0
+./ipvsadm --add-laddr -z 192.168.100.201 -t 123.1.2.3:80 -F dpdk0
+
+# add route to kni device.
+./dpip route add 172.10.0.2/30 dev dpdk1 scope kni_host
+```
+
+Then, the `zebra/ospfd` part. Firstly, to run the OSPF protocol between `DPVS` server and wan-side L3-switch, with the "inter-connection network" (here is `172.10.0.2/30`). For `DPVS`, we set the inter-connection IP on `dpdk1.kni`.
```
Comment (ywc689) [inline doc/tutorial.md]: to run -> run

File: doc/tutorial.md [hunk 14]
```diff
@@ -0,0 +1,594 @@
...
+
+```bash
+$ cat /etc/quagga/zebra.conf  # may installed to other path
+! -*- zebra -*-
+!
+! zebra sample configuration file
+!
+! Id: zebra.conf.sample,v 1.1 2002/12/13 20:15:30 paul Exp $
+!
+hostname localhost.localdomain
```
Comment (ywc689) [inline doc/tutorial.md]: it's better to configure real hostname

File: doc/tutorial.md [hunk 15]
```diff
@@ -0,0 +1,594 @@
...
+
+log file /var/log/quagga/zebra.log
+service password-encryption
+```
+
+For `ospfd`, these parameters need be set:
+
+* interface: it's WAN interface `dpdk1.kni`
+* route-id: not that significant, just use the LAN IP.
+* network: which network to broadcasting
```
Comment (ywc689) [inline doc/tutorial.md]: broadcasting -> broadcast or advertise

File: doc/tutorial.md [hunk 16]
```diff
@@ -0,0 +1,594 @@
...
+client: curl 123.1.2.3
+```
+
+<a id='fnat-keepalive'/>
+
+## Full-NAT with Keepalived (one-arm)
+
+This is an example for FullNAT used in internal network (LAN). `Keepalived`  (*`DPVS` modified version*) is used for to make DPVS works as *Master/Backup* model.
+
+![fnat-keepalive](pics/fnat-keepalived.png)
```
Comment (ywc689) [inline doc/tutorial.md]: diagram: the lines better to connect dpdk0.kni and VRRP switch?

File: doc/tutorial.md [hunk 17]
```diff
@@ -0,0 +1,594 @@
...
+}
+```
+
+Start `keepalived` on both Master and Backup.
+
+```bash
+./keepalived -f /etc/keepalived/keepalived.conf
+```
+
+For **test only**, add `VIP` and *routes* to DPDK interface manually on Master (do not set VIP on both master and backup), they should be added to keepalived configure file.
```
Comment (ywc689) [inline doc/tutorial.md]: vip can be added to dpdk interfaces automatically, do not need manual configuration

File: doc/tutorial.md [hunk 18]
```diff
@@ -0,0 +1,594 @@
...
+```
+
+Seems good, then try access the VIP from client.
+
+```bash
+client$ curl 192.168.100.254
+Your ip:port : 192.168.100.146:42394
+```
+
+> We just explain how DPVS works with keepalived, and not verify the master/backup feature provided by keepalived works. Pls refer LVS docs if needed.
```
Comment (ywc689) [inline doc/tutorial.md]: verify -> verify if

File: doc/tutorial.md [hunk 19]
```diff
@@ -0,0 +1,594 @@
...
+
+<a id='dr'/>
+
+# DR Mode (one-arm)
+
+Let's make a simple example for DR mode, some user may need it.
```
Comment (ywc689) [inline doc/tutorial.md]: some user -> some users

File: doc/tutorial.md [hunk 20]
```diff
@@ -0,0 +1,594 @@
...
+        TCP_CHECK { # health check
+            nb_sock_retry 2
+            connect_timeout 3
+            connect_port 80
+        }
+    }
+}
+```
+
+The keepalived config for backup is same with Master, except the `state` should be 'BACKUP', and `priority` should be lower.
```
Comment (ywc689) [inline doc/tutorial.md]: same with -> the same with    (somewhere else also exists)

File: doc/tutorial.md [hunk 21]
```diff
@@ -0,0 +1,594 @@
...
+
+# DR Mode (one-arm)
+
+Let's make a simple example for DR mode, some user may need it.
+
+![dr-one-arm](./pics/dr-one-arm.png)
+
+To use DR:
+
+* dpvs need a LAN IP first. (for one-arm, it must be different from VIP).
```
Comment (ywc689) [inline doc/tutorial.md]: need -> needs

File: doc/tutorial.md [hunk 22]
```diff
@@ -0,0 +1,594 @@
...
+* the `RS` and `DPVS` must in same sub-network (*on-link*).
+* On `RS`: `VIP` must be added to its *lo* interface.
+* On `RS`: `arp_ignore` must be set to *lo* interface.
+
+> `DPVS` need an *RS-faced* IP it self (here means "LAN-side" IP, it's not same conception as Local-IP (LIP) used by FNAT, just an normal IP address). Because `DPVS` need communicated with `RS`es. For *one-arm*, this LAN IP and VIP are on same DPDK interface. But they cannot be same, because `VIP` will also be set on `RS`es, if we do not use a separated LAN-IP, `RS`es will not reply the ARP request. Furthermore, the LAN-IP of `DPVS` must add **before** VIP.
```
Comment (ywc689) [inline doc/tutorial.md]: need -> needs
it self -> itself
same -> the same
an *RS-faced*  -> a *RS-faced*
an normal -> a normal

File: doc/tutorial.md [hunk 23]
```diff
@@ -0,0 +1,594 @@
...
+> For *tow-arm* DR, `DPVS` also need a LAN side IP to talk with LAN-side hosts, while VIP is configured on client-faced (WAN) interface.
+
+On `DPVS`, The `DR` configuration can be,
+
+```bash
+# on DPVS
+
+# add an LAN IP for DPVS, it must be different from VIP
```
Comment (ywc689) [inline doc/tutorial.md]: an LAN IP -> LAN IP

File: doc/tutorial.md [hunk 24]
```diff
@@ -0,0 +1,594 @@
...
+
+<a id='snat'/>
+
+# SNAT Mode (two-arm)
+
+`SNAT` mode can be used to let hosts in internal network without WAN IP (e.g., servers in IDC) to have Internet access.
+
+To configure `SNAT`,
+
+* WAN-side IP must configure with `sapool` option.
```
Comment (ywc689) [inline doc/tutorial.md]: must configure -> must be configured

File: doc/tutorial.md [hunk 25]
```diff
@@ -0,0 +1,594 @@
...
+* SNAT use "match" service instead of *<vip:vport>* for TCP/UDP,
```
Comment (ywc689) [inline doc/tutorial.md]: use -> uses

File: doc/tutorial.md [hunk 26]
```diff
@@ -0,0 +1,594 @@
...
+* default route may needed on DPVS WAN interface.
```
Comment (ywc689) [inline doc/tutorial.md]: may needed -> may be needed

File: doc/tutorial.md [hunk 27]
```diff
@@ -0,0 +1,594 @@
...
+
+> `match` support `proto`, `src-range`, `dst-range`, `oif` and `iif`. For example: `proto=tcp,src-range=192.168.0.0-192.168.0.254,dst-range=0.0.0.0:1-1024,oif=dpdk1`.
```
Comment (ywc689) [inline doc/tutorial.md]: support -> supports

File: doc/tutorial.md [hunk 28]
```diff
@@ -0,0 +1,594 @@
...
+# add WAN-side IP with sapool
+./dpip addr add $WAN_IP/$WAN_PREF dev dpdk1 sapool # must add sapool for WAN-side IP
+# add LAN-side IP as well as LAN route (generated)
+./dpip addr add $LAN_IP/$LAN_PREF dev dpdk0
+
+# add default route for WAN interface
+./dpip route add default via $GATEWAY dev dpdk1
+
+# SNAT section
+# -H MATCH       SNAT use -H for "match" service instead of -t or -u
```
Comment (ywc689) [inline doc/tutorial.md]: use -> uses

File: doc/tutorial.md [hunk 29]
```diff
@@ -0,0 +1,594 @@
...
+```bash
+host$ ping www.iqiyi.com
+host$ curl www.iqiyi.com
+```
+
+<a id='bond-vlan'/>
+
+# Bonding and VLAN devices
+
+`DPVS` support *Bonding* and *VLAN* virtual devices.
```
Comment (ywc689) [inline doc/tutorial.md]: support -> supports

File: doc/tutorial.md [hunk 30]
```diff
@@ -0,0 +1,594 @@
...
+
+For Bonding device, both `DPVS` and Switch need setting the Bonding interfaces with same Bonding mode. To enable Bonding device on `DPVS`, pls refer `conf/dpvs.bond.conf.sample`. Each Bonding device need one or more DPDK Physical device (`dpdk0`, ...) to work as it's slaves.
```
Comment (ywc689) [inline doc/tutorial.md]: need setting -> need to set

File: doc/tutorial.md [hunk 31]
```diff
@@ -0,0 +1,594 @@
...
+
+To use *VLAN* device, you can use `dpip` tool, *VLAN* device can be created based on real DPDK Physical device (e.g., `dpdk0`, `dpdk1`) or Bonding device (e.g., `bond0`). But cannot create VLAN device on VLAN device.
+
+This the the VLAN example, pls check `dpip vlan help` for more info.
```
Comment (ywc689) [inline doc/tutorial.md]: This the the -> This is the

File: doc/tutorial.md [hunk 32]
```diff
@@ -0,0 +1,594 @@
...
+```bash
+./dpip vlan add dpdk0.100 link dpdk0 proto 802.1q id 100
+./dpip vlan add link dpdk0 proto 802.1q id 101            # auto generate dev name
+./dpip vlan add link dpdk1 id 102
+./dpip vlan add link bond1 id 103
+```
+
+![bond-vlan-kni](./pics/bond-vlan-kni.png)
+
+Like DPDK Physical device, the *Bonding* and *VLAN* Virtual devices (e.g., `bond0` and `dpdk0.100`) has there own related `KNI` device on Linux environment (e.g., `bond0.kni`, `dpdk0.100.kni`).
```
Comment (ywc689) [inline doc/tutorial.md]: has there own related `KNI` device -> have there own related `KNI` devices

File: doc/tutorial.md [hunk 33]
```diff
@@ -0,0 +1,594 @@
...
+
+To configure `DPVS` (`FNAT`/`DR`/`SNAT`, `one-arm`/`two-arm`, `keepalived`/`ospfd`) for Virtual device is nothing special. Just "replace" the logical interfaces on sections above (like `dpdk0`, `dpdk1`, `dpdk1.kni`) to corresponding virtual devices.
```
Comment (ywc689) [inline doc/tutorial.md]: to corresponding virtual devices ->  with corresponding virtual devices

File: doc/tutorial.md (hunk 30 above)
Comment (ywc689) [reply doc/tutorial.md]: we just support bonding mode 0, 4 for now

File: doc/tutorial.md (hunk 16 above)
Comment (beacer) [reply doc/tutorial.md]: the line is for data forwarding.

File: doc/tutorial.md [hunk 34]
```diff
@@ -57,7 +57,7 @@ For *one-arm*, only one DPDK intreface needed, and you can refer `conf/dpvs.conf
...
 
-Consider `DPDK` application manages the networking interface completely (except the extra control NIC if exist), Linux Kernel and programs run on Kernel TCP/IP stack cannot receive packets from `DPDK` interface directly. To make Linux programs like `sshd`, `zebra/ospfd` and `keepalived` work, DPDK `kni` device is used. Then the Linux programs can working on `kni` device with Linux TCP/IP stack. Actually, `DPVS` pass the packets it's not interested in to `kni` device, like `OSPF/VRRP/ssh` packets. So that the programs "listening" on Linux stack are able to handle them.
+Considering `DPDK` application manages the networking interface completely (except the extra control NIC if exist), Linux Kernel and programs run on Kernel TCP/IP stack cannot receive packets from `DPDK` interface directly. To make Linux programs like `sshd`, `zebra/ospfd` and `keepalived` work, DPDK `kni` device is used. Then the Linux programs can working on `kni` device with Linux TCP/IP stack. Actually, `DPVS` pass the packets, whose it's not interested in, to `kni` device. For instance, *OSPF/VRRP/ssh* packets. So that the programs "working" on Linux stack are able to handle them.
```
Comment (ywc689) [inline doc/tutorial.md]: pass -> passes
whose -> which

File: doc/tutorial.md [hunk 35]
```diff
@@ -399,7 +394,7 @@ Start `keepalived` on both Master and Backup.
 ./keepalived -f /etc/keepalived/keepalived.conf
 ```
 
-For **test only**, add `VIP` and *routes* to DPDK interface manually on Master (do not set VIP on both master and backup), they should be added to keepalived configure file.
+For **test only**, add `VIP` and *routes* to DPDK interface manually on Master. Do not set VIP on both master and backup, on practice they should be added to keepalived configure file.
```
Comment (ywc689) [inline doc/tutorial.md]: on practice -> in practice

File: doc/tutorial.md [hunk 36]
```diff
@@ -444,32 +439,32 @@ client$ curl 192.168.100.254
...
+* dpvs needs a LAN IP first. (for one-arm, it must be different from VIP).
 * the `RS` and `DPVS` must in same sub-network (*on-link*).
 * On `RS`: `VIP` must be added to its *lo* interface.
 * On `RS`: `arp_ignore` must be set to *lo* interface.
 
-> `DPVS` need an *RS-faced* IP it self (here means "LAN-side" IP, it's not same conception as Local-IP (LIP) used by FNAT, just an normal IP address). Because `DPVS` need communicated with `RS`es. For *one-arm*, this LAN IP and VIP are on same DPDK interface. But they cannot be same, because `VIP` will also be set on `RS`es, if we do not use a separated LAN-IP, `RS`es will not reply the ARP request. Furthermore, the LAN-IP of `DPVS` must add **before** VIP.
+> `DPVS` needs a *RS-faced* IP itself (here means "LAN-side" IP, it's not the same conception as Local-IP (LIP) used by FNAT, just a normal IP address). Because `DPVS` need communicated with `RS`es. For *one-arm*, this LAN IP and VIP are on same DPDK interface. But they cannot be same, because `VIP` will also be set on `RS`es, if we do not use a separated LAN-IP, `RS`es will not reply the ARP request. Furthermore, the LAN-IP of `DPVS` must be add **before** VIP.
```
Comment (ywc689) [inline doc/tutorial.md]: be add -> be added

**Pull request merged**